proof/refine/ARM/IpcCancel_R.thy

```diff
@@ -2521,6 +2439,445 @@ lemma possibleSwitchTo_all_invs_but_ct_not_inQ':
   apply (intro conjI; fastforce?)
   by (metis fold_list_refs_of_replies')
 
+lemma possibleSwitchTo_sch_act_not_other:
+  "\<lbrace>\<lambda>s. sch_act_not t' s \<and> t' \<noteq> t\<rbrace>
+   possibleSwitchTo t
+   \<lbrace>\<lambda>_. sch_act_not t'\<rbrace>"
+  apply (clarsimp simp: possibleSwitchTo_def)
+  apply (wpsimp wp: threadGet_wp inReleaseQueue_wp)
+  apply (clarsimp simp: obj_at'_def)
+  done
+
+(* FIXME RT Michael: also in Victor's PR *)
+lemma replyUnlink_weak_sch_act_wf[wp]:
+  "\<lbrace>\<lambda>s. weak_sch_act_wf (ksSchedulerAction s) s \<and> sch_act_not t s\<rbrace>
+   replyUnlink r t
+   \<lbrace>\<lambda>_ s. weak_sch_act_wf (ksSchedulerAction s) s\<rbrace>"
+  unfolding replyUnlink_def setReplyTCB_def getReplyTCB_def
+  by (wpsimp wp: hoare_vcg_imp_lift hoare_vcg_all_lift gts_wp'
+           simp: weak_sch_act_wf_def)
+
+lemma restart_thread_if_no_fault_reply_tcb_reply_at[wp]:
+  "restart_thread_if_no_fault t  \<lbrace>reply_tcb_reply_at ((=) (Some t')) r_opt\<rbrace>"
+  apply (clarsimp simp: sk_obj_at_pred_def)
+  apply (wpsimp wp: restart_thread_if_no_fault_obj_at_impossible')
+  done
+
+lemma reply_unlink_tcb_reply_tcb_reply_at_other:
+  "\<lbrace>\<lambda>s. reply_tcb_reply_at ((=) (Some t')) r' s \<and> t' \<noteq> t\<rbrace>
+   reply_unlink_tcb t r
+   \<lbrace>\<lambda>_ s. reply_tcb_reply_at ((=) (Some t')) r' s\<rbrace>"
+  apply (wpsimp wp: update_sk_obj_ref_wps get_simple_ko_wp gts_wp
+              simp: reply_unlink_tcb_def)
+  apply (clarsimp simp: reply_tcb_reply_at_def obj_at_def pred_tcb_at_def)
+  done
+
+abbreviation (input)
+  "reply_unlink_ts_pred t s
+    \<equiv> \<forall>ep r_opt pl.
+       pred_map ((=) (Structures_A.thread_state.BlockedOnReceive ep (Some r_opt) pl)) (tcb_sts_of s) t
+       \<longrightarrow> reply_tcb_reply_at ((=) (Some t)) r_opt s"
+
+abbreviation (input)
+  "cancel_all_ipc_loop_body t
+    \<equiv> do st \<leftarrow> get_thread_state t;
+          reply_opt \<leftarrow> case st of Structures_A.thread_state.BlockedOnReceive x r_opt xa
+                                    \<Rightarrow> return r_opt
+                                | _ \<Rightarrow> return None;
+          y \<leftarrow> when (reply_opt \<noteq> None) $ reply_unlink_tcb t (the reply_opt);
+          restart_thread_if_no_fault t
+       od"
+
+lemma cancel_all_ipc_loop_body_reply_unlink_ts_pred_other:
```

Comment:
```suggestion
(* FIXME RT: move to AInvs *)
lemma cancel_all_ipc_loop_body_reply_unlink_ts_pred_other:
```

Reply:
I've gone with the FIXME RT: move to AInvs for the moment. I'm guessing some of these will be useful in the AInvs proofs, but it'd probably take a bit too much time to figure out exactly how